### PR TITLE
Lock writes to _overallBuildSuccess

### PR DESCRIPTION
**Pull request by cdmihai**
Hopefully fixes https://teams.microsoft.com/l/message/19:3212bf033f4c4b5198643a04fa1048fa@thread.skype/1620160142660?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=4ba7372f-2799-4677-89f0-7a1aaea3706c&parentMessageId=1620160142660&teamName=.NET%20Developer%20Experience&channelName=MSBuild&createdTime=1620160142660

### Changes Made
Protected writes to _overallBuildSuccess with _syncLock

### CODE REVIEW

File: src/Build/BackEnd/BuildManager/BuildManager.cs [hunk 1]
```diff
@@ -1628,12 +1628,16 @@ private void HandleExecuteSubmissionException(GraphBuildSubmission submission, E
...
-            if (submission.IsStarted)
+            lock (_syncLock)
             {
-                submission.CompleteResults(new GraphBuildResult(submission.SubmissionId, ex));
+                if (submission.IsStarted)
+                {
+                    submission.CompleteResults(new GraphBuildResult(submission.SubmissionId, ex));
+                }
+
+                _overallBuildSuccess = false;
```
Comment (ladipro) [inline src/Build/BackEnd/BuildManager/BuildManager.cs:1638]: There's a similar write at:

https://github.com/dotnet/msbuild/blob/b02f8de7e24d541df1d8c7d71cf8036d0ce91213/src/Build/BackEnd/BuildManager/BuildManager.cs#L1608

Why is it not protected by the lock as well?

Comment (BenVillalobos) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1638]: On line 857, 2403, and 2417. Though when they're called there might have been locked already.

Comment (cdmihai) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1641]: I tried to add the minimal number of locks (though I had missed the one @ladipro mentioned). If a write is not under a lock it means the method is called from other methods that do lock (or I missed them).  Do you prefer if I just lock every single write regardless if it's locked in a calling method?

Comment (BenVillalobos) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1641]: > Do you prefer if I just lock every single write regardless if it's locked in a calling method?
definitely not 🙂 I did a quick crtl+f and commented while realizing some may have already been locked

**Review by ladipro — APPROVED** [review]

**Review by rainersigwald — APPROVED** [review]

**Review by BenVillalobos — APPROVED** [review]

File: src/Build/BackEnd/BuildManager/BuildManager.cs (hunk 1 above)
Comment (danmoseley) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1641]: As a general practice, you might consider adding `Debug.Assert(Monitor.IsEntered(lockobj))` at the top of all methods that need to be called within a lock, so it's not necessary to mentally infer it.

Comment (danmoseley) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1641]: I've done this in other projects and it's been very handy to protect against future maintenance errors

Comment (cdmihai) [reply src/Build/BackEnd/BuildManager/BuildManager.cs:1650]: Well that's a cool trick to keep around! I added checks to non-locking but state mutating methods in the BuildManager and found a few that weren't protected.

**Review by rainersigwald — APPROVED** [review]

**Pull request merged**